File: html2uxml/resolver.py
"""Match CSS rules against the parsed HTML tree and produce per-element styles."""
from __future__ import annotations

from dataclasses import dataclass, field

from .css_parser import (
    Declaration,
    Rule,
    Selector,
    parse_selector,
)
from .html_parser import Node


@dataclass
class _ParsedRule:
    selectors: list[Selector]    # one entry per comma-separated selector
    declarations: list[Declaration]
    order: int


def parse_rules(rules: list[Rule]) -> list[_ParsedRule]:
    out: list[_ParsedRule] = []
    for r in rules:
        sels: list[Selector] = []
        for raw in r.selectors:
            s = parse_selector(raw)
            if s is not None:
                sels.append(s)
        if sels:
            out.append(_ParsedRule(selectors=sels, declarations=r.declarations, order=r.order))
    return out


# ---------------------------------------------------------------------------
# Selector matching
# ---------------------------------------------------------------------------


def _matches_compound(node: Node, comp) -> bool:
    if comp.tag != "*" and node.tag != comp.tag:
        return False
    if comp.id is not None and node.attrs.get("id") != comp.id:
        return False
    if comp.classes:
        node_classes = set(node.classes())
        for c in comp.classes:
            if c not in node_classes:
                return False
    for op, name, value in comp.attrs:
        actual = node.attrs.get(name)
        if op == "":
            if actual is None:
                return False
        elif op == "=":
            if actual != value:
                return False
        elif op == "~=":
            if actual is None or value not in actual.split():
                return False
        elif op == "|=":
            if actual is None or (actual != value and not actual.startswith(value + "-")):
                return False
        elif op == "^=":
            if actual is None or not actual.startswith(value):
                return False
        elif op == "$=":
            if actual is None or not actual.endswith(value):
                return False
        elif op == "*=":
            if actual is None or value not in actual:
                return False
    return True


def _matches_chain(node: Node, chain: list, parent_chain: list[Node],
                   sibling_index: int, siblings: list[Node]) -> bool:
    if not chain:
        return False
    combinator, compound = chain[-1]
    if not _matches_compound(node, compound):
        return False
    if len(chain) == 1:
        return True
    rest = chain[:-1]
    if combinator == ">":
        if not parent_chain:
            return False
        parent = parent_chain[-1]
        gs = [c for c in parent.children if not c.is_text]
        try:
            pidx = gs.index(parent)
        except ValueError:
            pidx = -1
        return _matches_chain(parent, rest, parent_chain[:-1], pidx, gs)
    if combinator == "+":
        if sibling_index <= 0:
            return False
        prev = siblings[sibling_index - 1]
        return _matches_chain(prev, rest, parent_chain, sibling_index - 1, siblings)
    if combinator == "~":
        for i in range(sibling_index - 1, -1, -1):
            if _matches_chain(siblings[i], rest, parent_chain, i, siblings):
                return True
        return False
    # Default: descendant. Try every ancestor.
    for i in range(len(parent_chain) - 1, -1, -1):
        # Compute that ancestor's siblings/index for further left walks.
        anc = parent_chain[i]
        if i > 0:
            grand = parent_chain[i - 1]
            sibs = [c for c in grand.children if not c.is_text]
            try:
                idx = sibs.index(anc)
            except ValueError:
                idx = -1
        else:
            sibs, idx = [anc], 0
        if _matches_chain(anc, rest, parent_chain[:i], idx, sibs):
            return True
    return False


def selector_matches(node: Node, selector: Selector,
                     ancestors: list[Node], sibling_index: int,
                     siblings: list[Node]) -> bool:
    return _matches_chain(node, selector.chain, ancestors, sibling_index, siblings)


# ---------------------------------------------------------------------------
# Cascade
# ---------------------------------------------------------------------------


@dataclass
class ResolvedStyle:
    """Final styles for one element."""
    base: list[tuple[str, str]] = field(default_factory=list)
    pseudo_rules: list[tuple[str, list[Declaration]]] = field(default_factory=list)
    matched_rule_indices: list[int] = field(default_factory=list)
    # Declarations from selectors that USS won't parse (attribute selectors,
    # ::pseudo-elements, :nth-child, etc). These are hoisted onto the
    # element's per-instance USS rule so the styles still apply.
    unsupported_decls: list[Declaration] = field(default_factory=list)
    # When ::before / ::after rules with `content:` match this element,
    # synthesize sibling Labels.
    before_content: str | None = None
    after_content: str | None = None
    before_decls: list[Declaration] = field(default_factory=list)
    after_decls: list[Declaration] = field(default_factory=list)


def resolve(root: Node, parsed_rules: list[_ParsedRule]) -> dict[int, ResolvedStyle]:
    """Walk the tree; return a map from id(node) to ResolvedStyle."""
    out: dict[int, ResolvedStyle] = {}
    _walk(root, [], parsed_rules, out)
    return out


def _walk(node: Node, ancestors: list[Node], rules: list[_ParsedRule],
          out: dict[int, ResolvedStyle]) -> None:
    if not node.is_text and node.tag != "__root__":
        # Compute siblings + sibling_index for use in selector matching.
        if ancestors:
            parent = ancestors[-1]
            sibs = [c for c in parent.children if not c.is_text]
            try:
                sib_idx = sibs.index(node)
            except ValueError:
                sib_idx = -1
        else:
            sibs, sib_idx = [node], 0
        out[id(node)] = _resolve_for(node, ancestors, rules, sib_idx, sibs)
    for child in node.children:
        if child.is_text:
            continue
        _walk(child, ancestors + [node], rules, out)


def _has_pseudo(selector: Selector) -> bool:
    return any(c.pseudo for _, c in selector.chain)


def _resolve_for(node: Node, ancestors: list[Node], rules: list[_ParsedRule],
                 sib_idx: int, siblings: list[Node]) -> ResolvedStyle:
    matches: list[tuple[tuple[int, int, int], int, list[Declaration]]] = []
    pseudo_rules: list[tuple[str, list[Declaration]]] = []
    matched_indices: list[int] = []
    unsupported_decls: list[Declaration] = []
    before_content: str | None = None
    after_content: str | None = None
    before_decls: list[Declaration] = []
    after_decls: list[Declaration] = []
    for idx, rule in enumerate(rules):
        for sel in rule.selectors:
            if not selector_matches(node, sel, ancestors, sib_idx, siblings):
                continue
            # Pseudo-element ::before / ::after rules become synthetic siblings.
            pseudo_elem = _pseudo_element_kind(sel)
            if pseudo_elem in ("before", "after"):
                content_value = None
                rest: list[Declaration] = []
                for d in rule.declarations:
                    if d.prop == "content":
                        content_value = _strip_quotes(d.value)
                    else:
                        rest.append(d)
                if pseudo_elem == "before":
                    if content_value is not None:
                        before_content = content_value
                    before_decls.extend(rest)
                else:
                    if content_value is not None:
                        after_content = content_value
                    after_decls.extend(rest)
                continue
            if sel.has_unsupported_features():
                unsupported_decls.extend(rule.declarations)
                continue
            if _has_pseudo(sel):
                pseudo_rules.append((sel.raw, rule.declarations))
            else:
                matches.append((sel.specificity(), rule.order, rule.declarations))
                if idx not in matched_indices:
                    matched_indices.append(idx)
    matches.sort(key=lambda t: (t[0], t[1]))

    # Apply: lowest specificity first, later wins. !important wins over normal.
    bag: dict[str, tuple[str, bool]] = {}  # prop -> (value, important)
    for _, _, decls in matches:
        for d in decls:
            cur = bag.get(d.prop)
            if cur is None or (d.important and not cur[1]) or (d.important == cur[1]):
                bag[d.prop] = (d.value, d.important)

    # Inline style="" attribute beats anything except !important rules.
    inline = node.attrs.get("style", "")
    if inline:
        from .css_parser import _parse_declarations
        for d in _parse_declarations(inline):
            cur = bag.get(d.prop)
            if cur is None or not cur[1] or d.important:
                bag[d.prop] = (d.value, d.important)

    base = [(prop, val) for prop, (val, _) in bag.items()]
    return ResolvedStyle(
        base=base, pseudo_rules=pseudo_rules,
        matched_rule_indices=matched_indices,
        unsupported_decls=unsupported_decls,
        before_content=before_content,
        after_content=after_content,
        before_decls=before_decls,
        after_decls=after_decls,
    )


def _pseudo_element_kind(sel) -> str | None:
    """If the rightmost compound has `::before` or `::after`, return that kind."""
    if not sel.chain:
        return None
    _, comp = sel.chain[-1]
    for ps in comp.pseudo:
        low = ps.lower()
        if low in ("::before", "::after"):
            return low.lstrip(":")
    return None


def _strip_quotes(s: str) -> str:
    s = s.strip()
    if len(s) >= 2 and s[0] == s[-1] and s[0] in ("'", '"'):
        return s[1:-1]
    return s

File: Editor/Converter/Resolver.cs
// source: html2uxml/resolver.py (cascade + tree walk half)
//
// CSS cascade: walks the HTML tree, matches every CssLoader.CssRule against
// every element, sorts the matches by specificity + source order, applies
// declarations with !important precedence, and finally lets inline
// `style="..."` overrides in (which beat normal stylesheet rules but lose
// to !important rules — see resolver._resolve_for in Python).
//
// Selector matching primitives live in ResolverMatching.cs.

using System;
using System.Collections.Generic;

namespace ODDGames.Html2Uxml.Editor.Converter
{
    public static class Resolver
    {
        // Public entry point.
        //
        // Returns a per-node ordered (prop, value) list that has already been
        // collapsed by the cascade — duplicates removed, !important honored,
        // inline style="" merged in. Callers are expected to feed this
        // straight into the USS emitter.
        //
        // The map is keyed by HtmlNode reference (object identity), matching
        // Python's `id(node)` keying.
        public static Dictionary<HtmlLoader.HtmlNode, List<KeyValuePair<string, string>>> Resolve(
            HtmlLoader.HtmlNode root,
            List<CssLoader.CssRule> rules)
        {
            var output = new Dictionary<HtmlLoader.HtmlNode, List<KeyValuePair<string, string>>>(
                ReferenceEqualityComparer<HtmlLoader.HtmlNode>.Instance);
            if (root == null) return output;
            Walk(root, new List<HtmlLoader.HtmlNode>(), rules, output);
            return output;
        }

        private static void Walk(
            HtmlLoader.HtmlNode node,
            List<HtmlLoader.HtmlNode> ancestors,
            List<CssLoader.CssRule> rules,
            Dictionary<HtmlLoader.HtmlNode, List<KeyValuePair<string, string>>> output)
        {
            if (!node.IsText && node.Tag != "__root__")
            {
                List<HtmlLoader.HtmlNode> sibs;
                int sibIdx;
                if (ancestors.Count > 0)
                {
                    var parent = ancestors[ancestors.Count - 1];
                    sibs = ResolverMatching.NonTextChildren(parent);
                    sibIdx = sibs.IndexOf(node);
                }
                else
                {
                    sibs = new List<HtmlLoader.HtmlNode> { node };
                    sibIdx = 0;
                }
                output[node] = ResolveFor(node, ancestors, rules, sibIdx, sibs);
            }

            // Append-then-recurse-then-pop avoids per-step list allocations
            // while preserving the same `ancestors + [node]` semantics.
            ancestors.Add(node);
            foreach (var child in node.Children)
            {
                if (child.IsText) continue;
                Walk(child, ancestors, rules, output);
            }
            ancestors.RemoveAt(ancestors.Count - 1);
        }

        // Per-element cascade.
        private static List<KeyValuePair<string, string>> ResolveFor(
            HtmlLoader.HtmlNode node,
            List<HtmlLoader.HtmlNode> ancestors,
            List<CssLoader.CssRule> rules,
            int sibIdx,
            List<HtmlLoader.HtmlNode> siblings)
        {
            // (specificity, order, decls) — sorted ascending so later wins.
            var matches = new List<(int Specificity, int Order, List<CssDeclaration> Decls)>();

            foreach (var rule in rules)
            {
                var sel = rule.ParsedSelector;
                if (sel == null) continue;
                if (!ResolverMatching.SelectorMatches(node, sel, ancestors, sibIdx, siblings))
                    continue;

                // ::before / ::after rules are intentionally skipped here:
                // the Python resolver materialises them as synthetic sibling
                // nodes with their own decl lists, but the spec's public C#
                // surface only exposes the per-element collapsed decl list.
                // Pseudo-element handling can layer back in later without
                // changing the cascade for the host element itself.
                if (PseudoElementKind(sel) != null) continue;

                // USS-unsupported features (attribute selectors, ::pseudo
                // elements, :nth-* on non-runtime pseudos, etc) are dropped
                // unless the rule also carries a runtime pseudo. This mirrors
                // _resolve_for's `unsupported_decls` branch — those decls
                // never make it into the base cascade.
                if (sel.HasUnsupportedFeatures())
                {
                    if (sel.HasRuntimePseudo()) continue; // runtime pseudo -> kept for USS emit, not for base cascade
                    continue;
                }

                // Rules whose chain includes a runtime pseudo (`:hover`,
                // `:focus`, ...) likewise don't contribute to the resting
                // base style — they only fire at runtime. The resolver in
                // Python keeps these on a `pseudo_rules` list; for the
                // collapsed-decl-list public surface we just exclude them.
                if (HasPseudo(sel)) continue;

                matches.Add((rule.Specificity, rule.OriginIndex, rule.RawDecls));
            }

            // Tuple compare: (specificity, order). Python's list.sort is
            // stable; OrderBy in C# is also stable.
            matches.Sort((a, b) =>
            {
                int c = a.Specificity.CompareTo(b.Specificity);
                return c != 0 ? c : a.Order.CompareTo(b.Order);
            });

            // bag: prop -> (value, important). Iteration order = insertion
            // order, which is what we emit at the end (Python uses dict[]
            // which preserves insertion order in 3.7+).
            var bag = new Dictionary<string, (string Value, bool Important)>(StringComparer.Ordinal);
            var insertionOrder = new List<string>();

            foreach (var (_, _, decls) in matches)
            {
                foreach (var d in decls)
                {
                    Apply(bag, insertionOrder, d.Prop, d.Value, d.Important);
                }
            }

            // Inline style="..." beats stylesheet rules (loses to !important).
            string inline = AttrUtil.Get(node.Attrs, "style");
            if (!string.IsNullOrEmpty(inline))
            {
                foreach (var d in CssParser.ParseDeclarations(inline))
                {
                    if (!bag.TryGetValue(d.Prop, out var cur))
                    {
                        Apply(bag, insertionOrder, d.Prop, d.Value, d.Important);
                    }
                    else if (!cur.Important || d.Important)
                    {
                        // Either the existing decl is non-important (inline
                        // wins) or the new inline decl is itself important
                        // (also wins).
                        bag[d.Prop] = (d.Value, d.Important);
                    }
                }
            }

            var output = new List<KeyValuePair<string, string>>(insertionOrder.Count);
            foreach (var prop in insertionOrder)
                output.Add(new KeyValuePair<string, string>(prop, bag[prop].Value));
            return output;
        }

        private static void Apply(
            Dictionary<string, (string Value, bool Important)> bag,
            List<string> insertionOrder,
            string prop, string value, bool important)
        {
            if (!bag.TryGetValue(prop, out var cur))
            {
                bag[prop] = (value, important);
                insertionOrder.Add(prop);
                return;
            }
            // Mirrors resolver._resolve_for's bag update predicate exactly:
            //   if cur is None or (d.important and not cur.important)
            //                   or (d.important == cur.important):
            //       bag[d.prop] = (d.value, d.important)
            // The third clause means "same importance => later wins".
            if ((important && !cur.Important) || important == cur.Important)
            {
                bag[prop] = (value, important);
            }
        }

        private static bool HasPseudo(CssSelector sel)
        {
            foreach (var (_, c) in sel.Chain)
                if (c.Pseudo.Count > 0) return true;
            return false;
        }

        private static string PseudoElementKind(CssSelector sel)
        {
            if (sel.Chain.Count == 0) return null;
            var (_, comp) = sel.Chain[sel.Chain.Count - 1];
            foreach (var ps in comp.Pseudo)
            {
                string low = ps.ToLowerInvariant();
                if (low == "::before") return "before";
                if (low == "::after") return "after";
            }
            return null;
        }
    }

    // Reference-equality comparer so the per-node dict keys mirror Python's
    // id(node) keying. Prevents two distinct nodes with structurally equal
    // contents from colliding (which would otherwise cascade-collapse the
    // wrong styles together).
    internal sealed class ReferenceEqualityComparer<T> : IEqualityComparer<T> where T : class
    {
        public static readonly ReferenceEqualityComparer<T> Instance = new ReferenceEqualityComparer<T>();
        public bool Equals(T x, T y) => ReferenceEquals(x, y);
        public int GetHashCode(T obj) => System.Runtime.CompilerServices.RuntimeHelpers.GetHashCode(obj);
    }
}
